### ISSUE TITLE
tuples containing #[must_use] types should be #[must_use]

### ISSUE THREAD

**Issue by cramertj** [rust-lang/rust#61061]
```rust
fn main() {
    Ok::<(), ()>(());
}
```
gives a warning, but
```rust
fn main() {
    (Ok::<(), ()>(()),);
}
```
does not.

**Comment by Centril** [comment]
@joshtriplett suggested that we test this out with crater to see if this would generate much substantial noise and how common it would be.

**Comment by cramertj** [comment]
I'd expect it'd be a pretty substantial source of warnings, but that's exactly why I think we should introduce it. People compiling with `#![deny(warnings)]` are *asking* for these sorts of bugs to be caught, and we should serve them appropriately ;)

In any case, running crater would require having an implementation to do a crater run on, so let's do it!

**Comment by scottmcm** [comment]
Let's do it.  (I'd personally even like an attribute to allow people to opt-in to this for `Option<T>`, for example, when T is itself `must_use`.)

**Comment by varkor** [comment]
I've opened a pull request to try this out: https://github.com/rust-lang/rust/pull/61100.

**Comment by Centril** [comment]
We discussed this on the language team meeting; folks felt generally positive, but we'd like to do a crater run to see what the impact would be like and if there would be many false positives.

**Comment by pickfire** [comment]
I tried this using `#[must_use]` but it does not seemed to complain when the part of `#[must_use]` is not used, is the the correct behavior?
```rust
struct A;

#[must_use]
struct B;

fn t() -> (A, B) {
    (A, B)
}

fn main() {
    let (a, _b) = t();
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7034a76d4d9d38154798fa3aec48f972

I hit this when doing https://docs.rs/hyper/0.14.7/hyper/client/conn/struct.Builder.html#method.handshake where I didn't use `connection` part of the tuple which cause it to not work, which I did not know is required to be used.

**Comment by scottmcm** [comment]
@pickfire putting it into a binding, even an unused one, is considered a use, like how `let _ = something_must_use();` doesn't lint.  With just calling `t()` it lints: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c2dc6ffbd470b5f098e4b1bdf653d57f
